core/src/main/java/org/apache/iceberg/avro/AvroIterable.java

```diff
@@ -97,10 +98,19 @@ public CloseableIterator<D> iterator() {
   }
 
   private DataFileReader<D> newFileReader() {
+    SeekableInput stream = null;
     try {
-      return (DataFileReader<D>)
-          DataFileReader.openReader(AvroIO.stream(file.newStream(), file.getLength()), reader);
+      stream = AvroIO.stream(file.newStream(), file.getLength());
+      return (DataFileReader<D>) DataFileReader.openReader(stream, reader);
```

Comment:
How about try-with resources ?

Reply:
it would be incorrect. DataFileReader controls the lifecycle of the stream and closes it once DataFileReader itself is closed. This change is about closing a stream when openReader throws